src/Permit2Relayer.sol

```diff
@@ -0,0 +1,144 @@
+        );
+
+        IERC20(permit.permitted.token).safeTransfer(recipient, permit.permitted.amount);
+    }
+
+    function requestCrossChainSwapPermit2(
+        address router,
+        address signer,
+        address tokenIn,
+        address tokenOut,
+        uint256 amount,
+        uint256 solverFee,
+        uint256 dstChainId,
+        address recipient,
+        IPermit2.PermitTransferFrom memory permit,
+        bytes calldata signature,
+        bytes calldata additionalData
+    ) external {
+        /// @notice Type string of the custom witness
+        string memory witnessTypeString = string(
+            abi.encodePacked(
+                WITNESS_TYPE_NAME,
+                "(address router,address tokenIn,address tokenOut,uint256 amount,uint256 solverFee,uint256 dstChainId,address recipient,bytes additionalData)"
+            )
+        );
+
+        /// @notice Type hash used to compute the witness hash
+        bytes32 witnessTypeHash = keccak256(bytes(witnessTypeString));
+
+        /// @notice The permit2 witnessTypeString parameter,
+        string memory permit2WitnessTypeString = string(
+            abi.encodePacked(
+                WITNESS_TYPE_NAME, " witness)", witnessTypeString, "TokenPermissions(address token,uint256 amount)"
+            )
+        );
+
+        IPermit2.SignatureTransferDetails memory transferDetails = ISignatureTransfer.SignatureTransferDetails({
+            to: address(this),
+            // we require the permit amount to be the same as the requested amount
+            requestedAmount: permit.permitted.amount
+        });
+
+        // By computing the witness here, we ensure that the permit was approved for that request id specifically.
+        // That same reasoning cannot be applied to the additionalData as it is controlled by the caller entirely.
+        bytes32 witness = keccak256(
+            abi.encode(
+                witnessTypeHash,
+                router,
+                tokenIn,
+                tokenOut,
+                amount,
+                solverFee,
+                dstChainId,
+                recipient,
+                keccak256(additionalData)
+            )
+        );
+
+        // Consume the permit
+        // No need to track used identifiers here as the Router will do that
+        // and this function only forwards tokens to it.
+        // The Router will then ensure that each request id is used at most once.
+        PERMIT2.permitWitnessTransferFrom(permit, transferDetails, signer, witness, permit2WitnessTypeString, signature);
+        // Forward the tokens to the Router
+        // The Router will handle the rest of the request logic
+        IERC20(permit.permitted.token).safeTransfer(router, permit.permitted.amount);
+    }
+
+    /// @notice Sets the Permit2 contract address
+    /// @dev TODO: This function is used for testing purposes only
+    function setPermit2Address(address permit2Address) external {
```

Comment:
without `onlyAdmin` I think this can this be called by anybody. I'm not sure I love it nonetheless, even for testing